File: src/vector.rs
use crate::util;

pub enum TupleType {
    Vector = 0,
    Point = 1,
}

pub type Vector = (f64, f64, f64, u8);
pub type Point = (f64, f64, f64, u8);

pub fn create_vector(x: f64, y: f64, z: f64) -> Vector {
    (x, y, z, TupleType::Vector as u8)
}

pub fn create_point(x: f64, y: f64, z: f64) -> Point {
    (x, y, z, TupleType::Point as u8)
}

pub fn cmp_vector(a: Vector, b: Vector) -> bool {
    if util::equal(a.0, b.0) == false {
        return false;
    }
    if util::equal(a.1, b.1) == false {
        return false;
    }
    if util::equal(a.2, b.2) == false {
        return false;
    }
    return true;
}

// TODO: operator overloading here

pub fn add(a: Vector, b: Vector) -> Vector {
    (a.0 + b.0, a.1 + b.1, a.2 + b.2, TupleType::Vector as u8)
}

pub fn diff(a: Vector, b: Vector) -> Vector {
    (a.0 - b.0, a.1 - b.1, a.2 - b.2, TupleType::Vector as u8)
}

pub fn negate(v: Vector) -> Vector {
    (-v.0, -v.1, -v.2, TupleType::Vector as u8)
}

pub fn scale(v: Vector, scalar: f64) -> Vector {
    (
        scalar * v.0,
        scalar * v.1,
        scalar * v.2,
        TupleType::Vector as u8,
    )
}

pub fn magnitude(v: Vector) -> f64 {
    f64::sqrt(f64::powi(v.0, 2) + f64::powi(v.1, 2) + f64::powi(v.2, 2))
}

pub fn normalize(v: Vector) -> Vector {
    let m = magnitude(v);
    (v.0 / m, v.1 / m, v.2 / m, TupleType::Vector as u8)
}

pub fn dot(a: Vector, b: Vector) -> f64 {
    a.0 * b.0 + a.1 * b.1 + a.2 * b.2
}

pub fn cross(a: Vector, b: Vector) -> Vector {
    // input order dependent
    (
        a.1 * b.2 - a.2 * b.1,
        a.2 * b.0 - a.0 * b.2,
        a.0 * b.1 - a.1 * b.0,
        TupleType::Vector as u8,
    )
}

File: src/vector.zig
const std = @import("std");
const math = std.math;

pub const TupleType = enum(u8) {
    vector,
    point,
};

pub fn Point(comptime T: type) type {
    return struct {
        x: T,
        y: T,
        z: T,
        _type: TupleType,

        pub fn init(x: T, y: T, z: T) Point(T) {
            return .{
                .x = x,
                .y = y,
                .z = z,
                ._type = TupleType.point,
            };
        }
    };
}

pub fn Vector(comptime T: type) type {
    return struct {
        x: T,
        y: T,
        z: T,
        _type: TupleType,

        pub fn init(x: T, y: T, z: T) Vector(T) {
            return .{
                .x = x,
                .y = y,
                .z = z,
                ._type = TupleType.vector,
            };
        }

        pub fn add(self: Vector(T), other: Vector(T)) Vector(T) {
            return .{
                .x = self.x + other.x,
                .y = self.y + other.y,
                .z = self.z + other.z,
                ._type = TupleType.vector,
            };
        }

        pub fn diff(self: Vector(T), other: Vector(T)) Vector(T) {
            return .{
                .x = self.x - other.x,
                .y = self.y - other.y,
                .z = self.z - other.z,
                ._type = TupleType.vector,
            };
        }

        pub fn negate(self: Vector(T)) Vector(T) {
            return .{
                .x = -self.x,
                .y = -self.y,
                .z = -self.z,
                ._type = TupleType.vector,
            };
        }

        pub fn scale(self: Vector(T), scalar: T) Vector(T) {
            return .{
                .x = self.x * scalar,
                .y = self.y * scalar,
                .z = self.z * scalar,
                ._type = TupleType.vector,
            };
        }

        pub fn magnitude(self: Vector(T)) T {
            return @sqrt(math.pow(T, self.x, 2) + math.pow(T, self.y, 2) + math.pow(T, self.z, 2));
        }

        pub fn normalize(self: Vector(T)) Vector(T) {
            const m = self.magnitude();
            return .{
                .x = self.x / m,
                .y = self.y / m,
                .z = self.z / m,
                ._type = TupleType.vector,
            };
        }

        pub fn dot(self: Vector(T), other: Vector(T)) T {
            return self.x * other.x + self.y * other.y + self.z * other.z;
        }

        pub fn cross(self: Vector(T), other: Vector(T)) Vector(T) {
            // input order dependent
            return .{
                .x = (self.y * other.z - self.z * other.y),
                .y = (self.z * other.x - self.x * other.z),
                .z = (self.x * other.y - self.y * other.x),
                ._type = TupleType.vector,
            };
        }
    };
}

test "Point.init" {
    const p = Point(f64).init(4.3, -4.2, 3.1);

    try std.testing.expectEqual(p, Point(f64).init(4.3, -4.2, 3.1));
}

test "Vector.init" {
    const v = Vector(f64).init(4.3, -4.2, 3.1);

    try std.testing.expectEqual(v, Vector(f64).init(4.3, -4.2, 3.1));
}

test "Vector.add" {
    const v1 = Vector(f64).init(3.0, -2.0, 5.0);
    const v2 = Vector(f64).init(-2.0, 3.0, 1.0);

    const expected = Vector(f64).init(1.0, 1.0, 6.0);
    const result = v1.add(v2);

    try std.testing.expectEqual(result, expected);
}

test "Vector.diff" {
    const v1 = Vector(f64).init(3.0, 2.0, 1.0);
    const v2 = Vector(f64).init(5.0, 6.0, 7.0);

    const expected = Vector(f64).init(-2.0, -4.0, -6.0);
    const result = v1.diff(v2);

    try std.testing.expectEqual(result, expected);
}

test "Vector.negate" {
    const v = Vector(f64).init(1.0, -2.0, 3.0);

    const expected = Vector(f64).init(-1.0, 2.0, -3.0);
    const result = v.negate();

    try std.testing.expectEqual(result, expected);
}

test "Vector.scale" {
    const v = Vector(f64).init(1.0, -2.0, 3.0);

    const expected = Vector(f64).init(3.5, -7.0, 10.5);
    const result = v.scale(3.5);

    try std.testing.expectEqual(result, expected);
}

test "Vector.magnitude" {
    var v = Vector(f64).init(1.0, 0.0, 0.0);
    try std.testing.expect(v.magnitude() == 1.0);

    v = Vector(f64).init(0.0, 1.0, 0.0);
    try std.testing.expect(v.magnitude() == 1.0);

    v = Vector(f64).init(0.0, 0.0, 1.0);
    try std.testing.expect(v.magnitude() == 1.0);

    v = Vector(f64).init(1.0, 2.0, 3.0);
    try std.testing.expect(v.magnitude() == @sqrt(14.0));

    v = Vector(f64).init(-1.0, -2.0, -3.0);
    try std.testing.expect(v.magnitude() == @sqrt(14.0));
}

// could break up into multiple tests
test "Vector.normalize" {
    var v = Vector(f64).init(4.0, 0.0, 0.0);
    var expected = Vector(f64).init(1.0, 0.0, 0.0);
    var result = v.normalize();

    try std.testing.expectEqual(result, expected);

    try std.testing.expect(result.magnitude() == 1.0);

    v = Vector(f64).init(1.0, 2.0, 3.0);
    const m = @sqrt(14.0);
    expected = Vector(f64).init(1.0 / m, 2.0 / m, 3.0 / m);
    result = v.normalize();

    try std.testing.expectEqual(result, expected);

    try std.testing.expect(result.magnitude() == 1.0);
}

test "Vector.dot" {
    const v1 = Vector(f64).init(1.0, 2.0, 3.0);
    const v2 = Vector(f64).init(2.0, 3.0, 4.0);

    try std.testing.expect(v1.dot(v2) == 20.0);
}

test "Vector.cross" {
    const v1 = Vector(f64).init(1.0, 2.0, 3.0);
    const v2 = Vector(f64).init(2.0, 3.0, 4.0);

    const expected = Vector(f64).init(-1.0, 2.0, -1.0);
    const result = v1.cross(v2);

    try std.testing.expectEqual(result, expected);
}
